GC USB Adapter configuration improvements

VinDuv opened the pull request:
Some improvements to the GameCube USB adapter configuration process:
 * If opening the adapter fails, unplugging and replugging it will make Dolphin try again to open it.
 * The adapter configuration dialog will display an error message (coming from libusb) if an error occurs when opening the adapter, instead of displaying “No Adapter Detected”.
 * The adapter configuration dialog will refresh if the adapter state changes. This was (partially I think) implemented in the Wx backend but not the Qt backend.

Caveats:
 * On systems without libusb hotplug, if the open process fails, it will be retried every half-second until it works or the controller is unplugged (On systems with libusb hotplug it will wait for it to be unplugged and replugged before trying again). Unplugging a successfully opened controller on these systems will not be detected until something reads from the controller. (The current system also has this issue)
* The error message from libusb is not translated. libusb has a couple of built-in translations, which can be activated by calling `libusb_setlocale` with a locale string. I did not implement this because I wasn’t sure where to put the call and how to get the current locale in a cross-platform manner.

BhaaLseN commented on Source/Core/Core/Analytics.cpp:
```diff
@@ -278,9 +278,9 @@ void DolphinAnalytics::MakePerGameBuilder()
+  builder.AddData("gcadapter-detected", GCAdapter::IsDetected(NULL));
   builder.AddData("has-controller", Pad::GetConfig()->IsControllerControlledByGamepadDevice(0) ||
-                                        GCAdapter::IsDetected());
+                                        GCAdapter::IsDetected(NULL));
```
This should probably use `nullptr` instead of `NULL`

VinDuv replied:
Right. Will fix.

BhaaLseN commented on Source/Core/DolphinQt/Config/Mapping/GCPadWiiUConfigDialog.cpp:
```diff
@@ -24,28 +24,30 @@ GCPadWiiUConfigDialog::GCPadWiiUConfigDialog(int port, QWidget* parent)
 
+  UpdateAdapterStatus();
+
+  GCAdapter::SetAdapterCallback([this] { QMetaObject::invokeMethod(this, "UpdateAdapterStatus"); });
```
Isn't there any other method to call `UpdateAdapterStatus` that doesn't rely on a magic string?

VinDuv replied:
I wasn’t aware of it, but Qt 5 apparently allows using member functions pointers for the signal/slots system instead of strings (https://wiki.qt.io/New_Signal_Slot_Syntax).
So `QMetaObject::invokeMethod(this, &GCPadWiiUConfigDialog::UpdateAdapterStatus);` actually works (and is apparently fully type-safe). I will change the code to use it.

BhaaLseN commented on Source/Core/InputCommon/GCAdapter.cpp:
```diff
@@ -115,11 +114,23 @@ static int HotplugCallback(libusb_context* ctx, libusb_device* dev, libusb_hotpl
       std::lock_guard<std::mutex> lk(s_init_mutex);
       AddGCAdapter(dev);
     }
+    else if ((s_status < 0) && (s_detect_callback != nullptr))
```
Those additional parentheses aren't really necessary, you can remove those (this also applies to pretty much all the other places)

BhaaLseN commented on Source/Core/InputCommon/GCAdapter.cpp:
```diff
@@ -28,7 +28,7 @@ static void ResetRumbleLockNeeded();
 static void Setup();
 
-static bool s_detected = false;
+static int s_status = 0;  // 0: No adapter detected; 1: Adapter detected; <0: libusb error code
```
Any thoughts on named constants (either `enum` or `constexpr`, whichever makes more sense) for the zero/one case?

BhaaLseN replied:
Looks better<i>ish</i> that way. You need not comment the _possible values_ like that though (at least not solely), it would be more useful if you could comment on its _purpose_.

lioncash commented on Source/Core/InputCommon/GCAdapter.cpp:
```diff
@@ -546,14 +575,20 @@ void Output(int chan, u8 rumble_command)
+  }
+
+  if (error_message)
+    *error_message = libusb_strerror((enum libusb_error)s_status);
```
`static_cast<libusb_error>(s_status)`

Techjar commented on Source/Core/DolphinQt/Config/Mapping/GCPadWiiUConfigDialog.cpp:
```diff
@@ -24,28 +24,31 @@ GCPadWiiUConfigDialog::GCPadWiiUConfigDialog(int port, QWidget* parent)
+  UpdateAdapterStatus();
+
+  GCAdapter::SetAdapterCallback(
+      [this] { QMetaObject::invokeMethod(this, &GCPadWiiUConfigDialog::UpdateAdapterStatus); });
```
You don't need `invokeMethod` or an anonymous function here, `&GCPadWiiUConfigDialog::UpdateAdapterStatus` can be passed directly to `SetAdapterCallback` as it has the appropriate params and return type.

Techjar commented on Source/Core/DolphinQt/Config/Mapping/GCPadWiiUConfigDialog.cpp:
```diff
@@ -24,28 +24,31 @@ GCPadWiiUConfigDialog::GCPadWiiUConfigDialog(int port, QWidget* parent)
+  UpdateAdapterStatus();
+
+  GCAdapter::SetAdapterCallback(
+      [this] { QMetaObject::invokeMethod(this, &GCPadWiiUConfigDialog::UpdateAdapterStatus); });
```
There's no need to use `QMetaObject::invokeMethod` here. You can simply put `UpdateAdapterStatus();` instead, as unqualified function names are implied to reference `this`.

VinDuv replied:
The callback will be called from a background thread. Is it safe to directly update the GUI from it?
(For the record, [the Wx code](https://github.com/dolphin-emu/dolphin/blob/227290055d6685281c72fd1d2076b67fdfb1b858/Source/Core/DolphinWX/Config/GCAdapterConfigDiag.cpp#L67) used `wxQueueEvent` to trigger the update on the main thread)

spycrab replied:
I'd suggest using ``QueueOnObject`` from our ``QtUtils/QueueOnObject.h`` header file instead.

VinDuv commented:
Updated PR:
* Added named constants for non-error adapter status
* Replaced call of `QMetaObject::invokeMethod` with `QueueOnObject` (should also fix build issues with earlier Qt versions)

VinDuv commented on Source/Core/InputCommon/GCAdapter.cpp:
```diff
@@ -28,7 +28,14 @@ static void ResetRumbleLockNeeded();
 static void Setup();
 
-static bool s_detected = false;
+enum
```
I don’t think I can use a typed enum because I put the status in an int variable (that can contain either one of the two enum values, or a (negative) libusb error code).

Maybe I could replace the enum with a struct wrapping the  status/error code, with `IsDetected`/`IsError`/`GetErrorCode`/`SetDetected`/`SetNotDetected`/`SetErrorCode` methods. What do you think?

leoetlino replied:
Oh, feel free to ignore this then.

VinDuv commented on Source/Core/InputCommon/GCAdapter.cpp:
```diff
@@ -203,6 +213,7 @@ void StartScanThread()
   if (ret < 0)
   {
     ERROR_LOG(SERIALINTERFACE, "libusb_init failed with error: %d", ret);
+    s_status = ret;
```
Thanks for the tip, it indeed simplifies things here.

VinDuv commented on Source/Core/DolphinQt/Config/Mapping/GCPadWiiUConfigDialog.h:
```diff
@@ -24,6 +25,10 @@ class GCPadWiiUConfigDialog final : public QDialog
   void CreateLayout();
   void ConnectWidgets();
 
+private slots:
```
Done.

VinDuv commented:
Fixed some issues and rebased this PR on top of PR #8070 as they had conflicting changes in `GCAdapter.cpp`.

leoetlino commented:
It looks like you forgot the libusb error enum change.

leoetlino commented:
Could you rebase this again please?

VinDuv commented:
Done. I’ve also removed the extraneous `enum` in the `static_cast`.
